dash_labs/dashdown.py

```diff
@@ -0,0 +1,419 @@
+"""
+ todo
+    - reduce security risk with exec
+       - local files only. Ensure that the file is within a certain directory, and by default
+       that should be the parent directory of the main app but maybe we could create a way to override that. Similar to /assets?
+    - css:
+       - create a MarkdownAIO stylesheet?
+       - eliminated dbc dependency  (replace Rows and Cols with inline css)
+       - document the default style in subcomponents and that user-supplied style will overwrite. Or better yet - use a stylesheet.
+    - rewrite the _remove_app_instance() to use the AST module
+      - allow for other names other than Dash ie app = DashProxy()
+    - any advantage to using jinja's read file function?
+    - see todos in pages.py in _register_page_from_markdown_file()
+    - use UUID for clipboard ID?
+    - add ability to change defaults "globally" so it doesn't have to be done for every instance.
+    - add ability to embed another file within the markdown file. Use case being, ability to keep the python app code in
+     a separate file so you can run it individually. We could integrate jinja in here perhaps…: {% include code.py %}
+    - Add markdown files to hot reload in dash.  That way users can have the same hot-reloading dev experience when working in markdown
+    - if code block is not executed don't register callbacks and layout
+    - Need to remove if __name__ == "__main__": ...   from the code blocks?
+    - refactor the _update_props function.
+
+"""
+
+from dash import dcc, html, dash_table, Input, Output, State, callback
+import re
+import ast
+import copy
+from jinja2 import Template
+import plotly.express as px
+import plotly
+import dash_bootstrap_components as dbc
+import warnings
+
+
+def warning_message(message, category, filename, lineno, line=None):
+    return f"{category.__name__}:\n {message} \n"
+
+
+warnings.formatwarning = warning_message
+
+
+class MarkdownAIO(html.Div):
+    def __init__(
+        self,
+        filename,
+        scope=None,
+        scope_creep=False,
+        dash_scope=True,
+        template_variables=None,
+        exec_code=False,
+        side_by_side=False,
+        code_first=True,
+        code_markdown_props={},
+        text_markdown_props={},
+        clipboard_props={},
+        app_div_props={},
+    ):
+        """
+        MarkdownAIO is an All-In-One component to display content of Markdown files with the option to run and/or
+         display code blocks.
+
+        - `filename`(string):
+           The path to the markdown file.
```

Comment:
Let's do relative to where it's called from. 

So these things should work:

`pages/content.md`
```
**custom content**
```
`pages/home.py`
```
dash.register_page(__name__)
layout = MarkdownAIO('pages/content.md')
```
`app.py`
```
dash.register_page('md', path='/example-2', layout=MarkdownAIO('pages/content.md')
```

And even this would work:
`tools/my_markdown_page.py`
```
def my_markdown_page(path):
    return MarkdownAIO(path, exec=True)
```
```
`app.py`
```
dash.register_page('md', path='/example-2', layout=tools.my_markdown_page('pages/content.md')
```

***

We'll need to update the `jinja` template to refer to the full path of the content too for consistency. So:
`pages/ex.md`
```
{% include "pages/example.py" %}
```

Reply:
Done